Meeting scheduler

NandhA-KuMar03 opened the pull request.

NandhA-KuMar03 committed:
completed excuses except update meeting

NandhA-KuMar03 committed:
Added update meeting controller service

NandhA-KuMar03 committed:
added unit tests and some minor changes

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/main/java/com/spring/rest/meetingscheduler/controller/MeetingController.java:
```diff
@@ -14,6 +18,7 @@
+import java.util.List;
 import java.util.Optional;
 
 @RestController
```
implement interfaces for methods

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/main/java/com/spring/rest/meetingscheduler/controller/MeetingController.java:
```diff
@@ -35,6 +41,7 @@ public HashMap<String,Integer> getAvailability(@RequestBody MeetingDetail meetin
     }
 
+//    Create meeting
     @PostMapping("/meeting")
```
use request mapping with method names for commonisation and check which to use and provide pointers on it

NandhA-KuMar03 replied:
Request mappings can be used for shared mappings incase we have same route names and different HTTP methods. For simpler controllers it is better to use specific HTTP annotations as it provides more readability. So I have used @GetMapping, @PostMapping instead of @RequestMappings

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/main/java/com/spring/rest/meetingscheduler/controller/MeetingController.java:
```diff
@@ -0,0 +1,76 @@
+        Time startTime = meetingDetail.getMeetingStartTime();
+        Time endTime = meetingDetail.getMeetingEndTime();
+        HashMap<String, Integer> map = meetingService.getAvailableRooms(date, startTime, endTime, count.orElse(0));
+        return map;
```
use service to generate meeting availability

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/main/java/com/spring/rest/meetingscheduler/entity/MeetingDetail.java:
```diff
@@ -0,0 +1,105 @@
+    @Column(name = "meeting_end_time")
+    private Time meetingEndTime;
+
+    @ManyToOne(cascade = {
```
give pointers why defaults are not used and whats the use of below mentioned cascades

NandhA-KuMar03 replied:
In default cascades all the cascades will be attached i.e., if we delete a row the corresponding value will be deleted in other table too. So in our case this meeting detail entity has employee ID. There is a specific use case that we should be able to remove people from the meeting. So in case we have default cascades if we remove people from it will also delete the employee. So to avoid that I have removed CascadeType.Remove and added others.

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/main/java/com/spring/rest/meetingscheduler/exception/RoomCapacityInsufficientException.java:
```diff
@@ -0,0 +1,15 @@
+package com.spring.rest.meetingscheduler.exception;
+
+public class RoomCapacityInsufficientException extends RuntimeException{
+    public RoomCapacityInsufficientException(String message) {
```
remove different exceptions to a common exception class and vary them with mesaages instead of different files

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/main/java/com/spring/rest/meetingscheduler/repository/EmployeeRepository.java:
```diff
@@ -0,0 +1,10 @@
+
+@Repository
+public interface EmployeeRepository extends JpaRepository<Employee, Integer> {
+    Employee findByEmployeeId(int id);
```
id should be maintained in long format research why and provide pointers

NandhA-KuMar03 replied:
Long is used for ID's since it provides a larger range when compared to int. Also since it is auto incremented, if the DB grows int cannot withstand the range of the values.

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/test/java/com/spring/rest/meetingscheduler/controller/MeetingControllerTest.java:
```diff
@@ -0,0 +1,113 @@
+    void getAvailability(){
+        //Given
+        HashMap<String, Integer> availabilityResponse = new HashMap<>();
+        MeetingDetail meetingDetailRequest = new MeetingDetail();
```
use common data instead of setting values

Naveen-nk-cdw commented on meeting_scheduler/meetingscheduler/src/test/java/com/spring/rest/meetingscheduler/controller/MeetingControllerTest.java:
```diff
@@ -0,0 +1,113 @@
+        assertEquals(response, expectedResponse);
+        assertEquals(OK, HttpStatus.valueOf(200));
+    }
+
```
cover exception and negative cases also

NandhA-KuMar03 committed:
Updated according to Review Comments

NandhA-KuMar03 committed:
Review Comments

Changed Controller names based on Rest practices
Changed Request Param Optional to required=false
written repos for CRUD operations
added constants file
Checked if the input value are of correct type eg.date and time
Changed response objects for certain routes
Changed respective tests too

NandhA-KuMar03 committed:
Review comments 2

Added everything in request body and not using both request body and request params.
Added new endpoints for CRUD operations - meeting room
Get meetings on specific date
Changed everything to response entity
changed respective test files in service
and started using mockMVC for controller tests

NandhA-KuMar03 committed:
Review comments 3

Changes to request structure and status codes
Add employee to team - bulk update
Added RequestPart file in create employee endpoint
Added tests to get coverage to 80%
